src/main/scala/org/broadinstitute/dsde/firecloud/service/UserService.scala

```diff
@@ -110,4 +123,100 @@ class UserService(rawlsDAO: RawlsDAO, thurloeDAO: ThurloeDAO, userToken: UserInf
+        "could not save Anonymous Group", exception))
+      case Success(_) => {
+        val futureAllKeys: Future[ProfileWrapper] = getAllKeysFromThurloe(userToken)
+        futureAllKeys map { keys => RequestComplete(keys)}
+      }
+    }
+  }
+
+  private def getAnonymousGroup: String = {
+    // randomly generate the anonymousGroupName, which follows format: terra-user-adjective-noun-endOfUUID
+    val anonymousGroupUUID: UUID = UUID.randomUUID()
+    val anonymousGroupName: String = (FireCloudConfig.FireCloud.supportPrefix
+      + getWord(anonymousGroupUUID.getMostSignificantBits(), UserService.randomAdjectiveList) + "-"
+      + getWord(anonymousGroupUUID.getLeastSignificantBits(), UserService.randomNounList) + "-"
+      + anonymousGroupUUID.toString().split("-")(4))
+    anonymousGroupName
+  }
+
+  private def getAllKeysFromThurloe(userToken: UserInfo): Future[ProfileWrapper] = {
+    thurloeDAO.getAllKVPs(userToken.id, userToken) map { // .getAllKVPs returns Option[ProfileWrapper]
+      case None => ProfileWrapper(userToken.id, List())
+      case Some(wrapper) => {
+        wrapper
+      }
+    }
+  }
+
+  /**
+    * creates a new anonymized Google group for the user
+    * @param keys the user's KVPs
+    * @param anonymousGroupName sets the name of the Google group to be created
+    * @return Future[PerRequestMessage] of all KVPs for the user
+    */
+  private def callCreateGoogleGroup(keys: ProfileWrapper, anonymousGroupName: String): Future[PerRequestMessage] = {
+    // define userEmail to add to google Group - check first for contactEmail, otherwise use user's login email
+    val userEmail = getProfileValue(keys, UserService.ContactEmailKey) match {
+      case None | Some("") => userToken.userEmail
+      case Some(contactEmail) => contactEmail // if there is a non-empty value set for contactEmail, we assume contactEmail is a valid email
+    }
+
+    googleServicesDAO.createGoogleGroup(anonymousGroupName, userEmail) match { // returns "" if group creation is not successful
+      case None | Some("") => {
+        Future(RequestComplete(keys))
+      }
+      case Some(groupEmailName) => {
+        writeAnonymousGroup(userToken, groupEmailName) // write new KVP to Thurloe
+      } // this returns a Future[PerRequestMessage]
+    }
+  }
+
+  /**
+    * gets all KVPs for the user from Thurloe
+    *   - checks whether the key `anonymousGroup` exists
+    *     - if `anonymousGroup` KVP exists for user:
+    *       - double check that that google group actually exists (and try to create it if it doesn't)
+    *       - return all keys
+    *     - if that key does not exist, we:
+    *       - generate an anonymized Google group for the user
+    *       - set that as the value for `anonymousGroup` KVP
+    *       - return all keys, including new `anonymousGroup`
+    * @param userToken
+    * @return
+    */
+  def getAllUserKeys(userToken: UserInfo): Future[PerRequestMessage] = {
```

Comment:
how often will this be called by Terra UI once it's fully integrated? my guess is once per full page reload since the keys should be stored in browser state. does that sound right? i'm asking because i'm trying to think about how much added volume to google group APIs this will be adding, but i think it's likely a non-issue.

this also feels a bit side-effecty to me. is there another trigger that could be used for creating the anonymized group? user registration seems like a good choice, but obviously that doesn't handle existing users. i'm not sure there's a better alternative, just thinking out loud

Reply:
yes, this should be called once per full page reload. we weren't sure how to evaluate whether this would add too much volume to google group APIs?

we'd discussed your suggested alternative - putting this in user registration and migrating existing users with a one-time script - but we landed on this route and folks seemed okay with it. what do you think, @MatthewBemis ?

Reply:
if it's once per full page reload then it's of no concern IMO. I just wanted to make sure that was the case since I didn't have all of the context. 

i think this approach is good- a migration script is much more difficult and not really worthwhile. this is essentially a runtime migration anyway :+1: